ext/ddtrace_profiling_native_extension/linux_tid_from_pthread.c

```diff
@@ -0,0 +1,209 @@
+    if (!reference_threads_probe_info[0].buffer_read_success) return -1;
+
+    // ...and then we create a few more to act as extra references
+    for (int i = 1; i < REFERENCE_THREADS_COUNT; i++) {
+      pthread_t reference_thread;
+      error = pthread_create(&reference_thread, NULL, collect_tid_probe_info, &reference_threads_probe_info[i]);
+      if (error) return -1;
+      error = pthread_join(reference_thread, NULL);
+      if (error || !reference_threads_probe_info[i].buffer_read_success) return -1;
+    }
+
+    short candidate_tid_offset = 0;
+
+    // We try to find an offset that contains the expected tid for all reference threads. If we get to the end of the
+    // buffer before finding such an offset, the `thread_offset` will be -1 and we'll report back an error.
+    while (candidate_tid_offset >= 0) {
+      bool all_offsets_match = true;
+
+      for (int i = 0; i < REFERENCE_THREADS_COUNT; i++) {
+        short thread_offset = find_tid_offset_in_buffer(&reference_threads_probe_info[i], candidate_tid_offset);
+
+        if (thread_offset != candidate_tid_offset) {
+          all_offsets_match = false;
+          candidate_tid_offset = thread_offset;
+        }
+      }
+
+      if (all_offsets_match) break;
+    }
+
+    return candidate_tid_offset;
+  }
+
+  pid_t linux_tid_from(pthread_t thread, short offset) {
+    char buffer[STRUCT_PTHREAD_READ_SIZE];
+    read_struct_pthread(thread, buffer);
+    return get_tid_from_buffer(buffer, offset);
+  }
+
+  static void *collect_tid_probe_info(void *tid_probe_info) {
+    struct tid_probe_info *result = (struct tid_probe_info *) tid_probe_info;
+
+    result->tid = ddtrace_gettid();
+    result->buffer_read_success = read_struct_pthread(pthread_self(), result->buffer);
+
+    return NULL;
+  }
+
+  bool read_safely(void *read_from_ptr, void *read_into_buffer, short buffer_size) {
+    struct iovec read_into = {.iov_base = read_into_buffer, .iov_len = buffer_size};
+    struct iovec read_from = {.iov_base = read_from_ptr, .iov_len = buffer_size};
+
+    int unused_flags = 0;
+    int number_of_iovecs = 1;
+
+    short num_read = process_vm_readv(ddtrace_gettid(), &read_into, number_of_iovecs, &read_from, number_of_iovecs, unused_flags);
+
+    return num_read == buffer_size;
+  }
+
+  static bool read_struct_pthread(pthread_t thread, void *struct_pthread_buffer) {
+    return read_safely((void *) thread, struct_pthread_buffer, STRUCT_PTHREAD_READ_SIZE);
+  }
+
+  static short find_tid_offset_in_buffer(struct tid_probe_info *tid_probe_info, short start_position) {
+    if (!tid_probe_info->buffer_read_success) return -1;
+
+    short buffer_size = STRUCT_PTHREAD_READ_SIZE;
+    pid_t *struct_pthread_base = (pid_t *) tid_probe_info->buffer;
+    pid_t reference_tid = tid_probe_info->tid;
+
+    for (pid_t *tid_ptr = struct_pthread_base + start_position; tid_ptr < struct_pthread_base + (buffer_size / sizeof(pid_t)); tid_ptr++) {
```

Comment:
Is it safe to assume that the tread_id we are looking for is of size `sizeof(pid_t)`?

Reply:
I believe so, for two kinds of reasons:

1. This comes from the linux kernel:

* It is a `pid_t` on both glibc ([link](https://github.com/bminor/glibc/blob/4290aed05135ae4c0272006442d147f2155e70d7/nptl/descr.h#L174))
* And an `int` on musl ([link](https://github.com/bminor/musl/blob/718f363bc2067b6487900eddc9180c84e7739f80/src/internal/pthread_impl.h#L35)) ([And then musl typedefs `int` as `pid_t`](https://github.com/bminor/musl/blob/718f363bc2067b6487900eddc9180c84e7739f80/include/alltypes.h.in#L46)
* The Linux kernel also uses `int` [as a typedef for it](https://github.com/torvalds/linux/blob/89bf6209cad66214d3774dac86b6bbf2aec6a30d/include/uapi/asm-generic/posix_types.h#L28).

2. We're limited by our read of exactly `STRUCT_PTHREAD_READ_SIZE` bytes. If somehow the `pid_t` info is not there, or the size/alignment changes so we don't find it, the code correctly handles this by reporting `-1` and thus we disable this approach.